paddle/fluid/imperative/utils/tensor_utils.cpp

```diff
@@ -0,0 +1,211 @@
+  RegisterNumpyType(map, NPY_BOOL, framework::proto::VarType::BOOL);
+  RegisterNumpyType(map, NPY_LONGLONG, framework::proto::VarType::SIZE_T);
+  RegisterNumpyType(map, NPY_INT16, framework::proto::VarType::INT16);
+  RegisterNumpyType(map, NPY_UINT8, framework::proto::VarType::UINT8);
+
+  return map;
+}
+
+static void CheckTensor(const framework::Tensor& tensor) {
+  // NOTE(minqiyang): add API hint when any check meets error here
+  // check if tensor is initialized
+  PADDLE_ENFORCE(tensor.IsInitialized(), "Tensor should be initialized first");
+
+  // check the place of tensor
+  // NOTE(minqiyang): we could only support numpy bridge
+  // between cpu tensor and ndarray now.
+  PADDLE_ENFORCE(paddle::platform::is_cpu_place(tensor.place()),
+                 "Can only convert CPU tensor to numpy.");
+}
+
+static int GetDType(const framework::Tensor& tensor) {
+  auto it = gNumpyDataTypeMap().proto_to_dtype_.find(
+      static_cast<int>(framework::ToDataType(tensor.type())));
+  if (it != gNumpyDataTypeMap().proto_to_dtype_.end()) {
+    return it->second;
+  }
+
+  PADDLE_THROW("Numpy conversion from tensor with type %s is NOT supported",
+               tensor.type().name());
+}
+
+static std::type_index GetTensorType(int dtype) {
+  auto it = gNumpyDataTypeMap().dtype_to_proto_.find(dtype);
+  if (it != gNumpyDataTypeMap().dtype_to_proto_.end()) {
+    return framework::ToTypeIndex(
+        static_cast<framework::proto::VarType::Type>(it->second));
+  }
+
+  PADDLE_THROW(
+      "Numpy conversion from numpy array with type %s is NOT supported", dtype);
+}
+
+static std::vector<npy_intp> DDimToNumpyShape(const framework::DDim& ddim) {
+  std::vector<npy_intp> numpy_shape;
+  numpy_shape.reserve(ddim.size());
+  for (size_t i = 0u; i != ddim.size(); ++i) {
+    numpy_shape.emplace_back(static_cast<npy_intp>(ddim[i]));
+  }
+  return numpy_shape;
+}
+
+static std::vector<int64_t> NumpyShapeToVector(int ndim,
+                                               npy_intp* numpy_shape) {
+  std::vector<int64_t> dims;
+  dims.reserve(ndim);
+  for (int i = 0; i != ndim; ++i) {
+    dims.emplace_back(numpy_shape[i]);
+  }
+
+  return dims;
+}
+
+// NOTICE(minqiyang): try pybind11 here and decide the way of
+// implementing numpy bridge after the benchmark.
+PyObject* TensorToNumpy(const framework::Tensor& tensor) {
+  // check the place of tensor
+  CheckTensor(tensor);
+
+  // get meta data (type, shape e.g..) from tensor
+  auto dtype = GetDType(tensor);
+  auto sizes = DDimToNumpyShape(tensor.dims());
+  // strides should take the sizeof type in count
+  auto strides = DDimToNumpyShape(framework::stride(tensor.dims()));
+  size_t type_size = framework::SizeOfType(tensor.type());
+  for (auto& stride : strides) {
+    stride *= type_size;
+  }
+
+  // create numpy array
+  PyObject* array = PyArray_New(
+      &PyArray_Type, tensor.dims().size(), sizes.data(), dtype, strides.data(),
+      tensor.Holder()->ptr(), 0, NPY_ARRAY_BEHAVED, nullptr);
```

Comment:
tensor.data<void>() or tensor.mutable_data<void>() is better.

Reply:
`strides` may be `nullptr` according the documentation https://docs.scipy.org/doc/numpy-1.13.0/reference/c-api.array.html

Reply:
I have a question about this, ```data()``` is a template function of type T and ```mutable_data()``` is not a const function.